proposals/21-11-transformer-annotations.md

```diff
@@ -0,0 +1,253 @@
+would like them to be verbose and communicate as much information as possible to the user.
+
+Kustomize previously had a [feature request](https://github.com/kubernetes-sigs/kustomize/issues/3979) to add an 
+option to kustomize build to retain the origin of a resource in an annotation, and a final design was settled 
+on after some discussion. It was implemented by [this pull request](https://github.com/kubernetes-sigs/kustomize/pull/4065) 
+and released in kustomize v4.3.0. This feature added a new field to the kustomization file, `buildMetadata`, an 
+array of boolean options such as `originAnnotations`. If `originAnnotations` is set, `kustomize build` will add annotations 
+to each resource describing the resource's origin. 
+
+While the above feature gives us valuable information about where many resources originated from, there are two 
+pieces missing:
+
+- Origin annotations for generated resources. A comment in the original issue proposed an annotation for generated 
+resources. Though this was accepted as part of the proposal, it was never implemented. However, the original proposal 
+for generated resources doesn't account for resources generated from remote bases.
+
+- Everything in kustomize is a transformer, so we would like to have annotations about which transformers, builtin or 
+custom, touched each resource. Builtin transformers include those that are invoked through various kustomization 
+fields. Some of these transformers are: 
+  - AnnotationsTransformer
+  - HashTransformer
+  - ImageTagTransformer
+  - LabelTransformer
+  - LegacyOrderTransformer
+  - NamespaceTransformer
+  - PatchJson6902Transformer
+  - PatchStrategicMergeTransformer
+  - PatchTransformer
+  - PrefixSuffixTransformer
+  - ReplacementTransformer
+  - ReplicaCountTransformer 
+  - ValueAddTransformer
+  
+  Custom transformers are invoked via the `transformers` field in the kustomization file.
+
+  For a user attempting to debug a large stream of output from `kustomize build`, knowing how each resource has been 
+  transformed through various layers will make it much easier to understand how the output was rendered, easing debugging 
+  and making changes to kustomization files.
+
+**Goals:**
+1. When `originAnnotations` is set, add annotations to generated resources that describe the source of the generator that 
+created them.
+2. Add a new option, `transformerAnnotations`, to `buildMetadata` that, when set, will annotate each resource with 
+information about which transformers, builtin or custom, touched each resource.
+
+**Non-goals:**
+1. Change the syntax of the existing `originAnnotations` option.
+2. Add these annotations to the output of kustomize by default.
+
+## Proposal
+
+### Generated Resources
+
+We should add the same set of annotations to generated resources as other resources when the `buildMetadata.originAnnotations` 
+option is set. If the source of the file points to a generator, we know that it is a generated resource.
```

Comment:
1. Yes. We can state this clearly in the docs.
2. That's a good point that I hadn't considered. If the information is available, I agree that generators like ConfigMapGenerator and SecretGenerator should get `kind` and `apiVersion` like the transformers. Considering `resources` as a generator, I can see how the `origin` annotation is providing data about where the generator is located and the new annotations for generated resources are consistent with that. I think having the same possible fields for both annotations is a great idea; even custom generators will have a `kind` and `apiVersion` that would be worth noting in the annotation. 
3. IMO it would be worth having a ConfigMapGenerator with `merge` set to true be a special case where we add a transformer annotation.

Reply:
> Let's aim for this then.

Sounds good, will update the proposal with this.

> One inconsistency to resolve is the meaning of "path" for builtin and external generators: right now, the former would use the path of the resource file itself, whereas the latter would point to the generator config. Maybe these should be two different fields? The path would be reserved for the specific file, and extensions generators could help us populate it via internal.config.kubernetes.io/path or something?

I suppose there is a difference there, and maybe we can have different names for the fields, e.g. `path` for the path to a resource file, vs `generatorConfig` for the path to the config of a generated resource. 

However, if a resource is generated, I'm not sure what it means to provide the path to the resource file itself. If someone has a generator config in their kustomization file, the generated resource gets added to the stream of yaml that is outputted to stdout, so at no point is there a path to a file containing the generated resource. In this case, would `path` need to be populated?

Reply:
>  If someone has a generator config in their kustomization file, the generated resource gets added to the stream of yaml that is outputted to stdout, so at no point is there a path to a file containing the generated resource. In this case, would path need to be populated?

In the case of a generator like ConfigMapGenerator I agree--path would not be populated as there's no sensible value for it. That's not necessarily true of all generated resources though: Imagine we make a ResourceGenerator implementation to back the resources field, and it is used from a generators or transformers field. In that case you have both fields, one pointing to the Kustomization or generator config file, and the other pointing to the resource file path. Third party generators would need to communicate the path back to us somehow, since even if the config includes a list of paths, we have no idea which resulting resources came from which files, if any.

> Is it necessary to put this in a single field? I'm thinking that if the transformer config has a name field, we can include that, but the format for stringified ResIds like "~G_~V_~K|name" that we currently have might be confusing to unfamiliar users.

Note that they can have a namespace, too. I'm not feeling strongly about this yet, but my thinking was that separate GVKNN fields might be more verbose than is warranted, depending on what we decide the target use case is (e.g. do we expect the user to want to parse out the GVKNN elements?). 

I agree that the stringified format is not very clear, nor consistent with how I've seen folks summarize this information in docs. I imagine this is already showing up and causing confusion. If it's only in error messages, I would love to improve it regardless of whether we use it here. I don't know whether it is ever encoded in results as of now.

Reply:
> In the case of a generator like ConfigMapGenerator I agree--path would not be populated as there's no sensible value for it. That's not necessarily true of all generated resources though: Imagine we make a ResourceGenerator implementation to back the resources field, and it is used from a generators or transformers field. In that case you have both fields, one pointing to the Kustomization or generator config file, and the other pointing to the resource file path. Third party generators would need to communicate the path back to us somehow, since even if the config includes a list of paths, we have no idea which resulting resources came from which files, if any.

Makes sense to me. If a generator creates a resource with `internal.config.kubernetes.io/path`, I'm fine with using that to populate the `config.kubernetes.io/origin` path field.

 > Note that they can have a namespace, too. I'm not feeling strongly about this yet, but my thinking was that separate GVKNN fields might be more verbose than is warranted, depending on what we decide the target use case is (e.g. do we expect the user to want to parse out the GVKNN elements?).

Wouldn't having it in the same field be almost equally verbose and potentially a bit more confusing and difficult to parse? The intended use case is that a user can look at these annotations and understand how the final output was rendered - this could arguably be achieved by just having the path to the generator or transformer config, but the apiVersion and kind will give more information about what kind of generation/transformation was done. I'm not necessarily against including the name/namespace in this output, but I'm not sure what additional value it will provide.

Reply:
>  I'm not necessarily against including the name/namespace in this output, but I'm not sure what additional value it will provide.

It's true that in most cases those values will be missing anyway (or defaulted), but e.g. if we're dealing with a Composition, or transformers inlined in the transformers field, there could be more than one of the same GVK in the same file that would be disambiguated by the name. Maybe too much of an edge case, but also probably not too difficult to accommodate if we omit when empty?

Reply:
> Maybe too much of an edge case, but also probably not too difficult to accommodate if we omit when empty?

If we're omitting when empty, I don't feel very strongly either way, so we can include it if you think it will be helpful. How do you envision it looking if it is in a single field vs separate fields? I'm inclined to support the latter for the sake of readability but could be convinced otherwise if you have an alternative, equally readable proposal.

Reply:
In typing out a response, I realized that what we have is an [ObjectReference](https://github.com/kubernetes/community/blob/64e708da7c6215d37c3b467a5d95b50748295672/contributors/devel/api-conventions.md#references-to-related-objects), so we should stick to the structure of that type, and probably isolate those fields within the origin/transformer data. I don't love my field names, but something like this:

```yaml
config.kubernetes.io/origin: |
   - path: path/to/deployment.yaml
     ref: v1.0.1
     repo: https://github.com/kubernetes-sigs/kustomize
     configuredIn: path/to/Kustomization.yaml
     configuredBy: 
       kind: ResourceGenerator
       apiVersion: builtin
       name: foo
       namespace: default
```

Reply:
I actually really like the names `configuredIn` and `configuredBy`. I will include it in my proposal.